fairseq/criterions/label_smoothed_cross_entropy.py

```diff
@@ -61,23 +66,32 @@ def __init__(
         label_smoothing,
         ignore_prefix_size=0,
         report_accuracy=False,
+        rdrop_alpha=0.0,
     ):
         super().__init__(task)
         self.sentence_avg = sentence_avg
         self.eps = label_smoothing
         self.ignore_prefix_size = ignore_prefix_size
         self.report_accuracy = report_accuracy
+        self.rdrop_alpha = rdrop_alpha
 
-    def forward(self, model, sample, reduce=True):
+    def forward(self, model, sample, reduce=True, net_output=None):
         """Compute the loss for the given sample.
 
         Returns a tuple with three elements:
         1) the loss
         2) the sample size, which is used as the denominator for the gradient
         3) logging outputs to display while training
         """
-        net_output = model(**sample["net_input"])
-        loss, nll_loss = self.compute_loss(model, net_output, sample, reduce=reduce)
+        if net_output is None:
+            if self.rdrop_alpha > 0 and sample["net_input"]["src_tokens"].size(
```

Comment:
@yuntang When do we use "source" or "features"?

Reply:
The outputs from HubertDataset and RawAudioDataset are "source" instead of "src_tokens". 
Also, we can find "source", "src_tokens" and "features" are all legal input types in sequence_generator
https://github.com/facebookresearch/fairseq/blob/main/fairseq/sequence_generator.py#L222-L241
So I guess we might meet "source" and "features" in the cost function too.

Reply:
@yuntang, I think both HubertDataset and RawAudioDataset are used during pre-training only. So we never use "source" or "features" during fine-tuning.